File: app/lib/features/trainer/domain/calendar_logic.dart
import 'calendar_assignment.dart';
import 'calendar_session.dart';

/// Solo la fecha, sin hora — para poder comparar días sin que la hora de
/// creación del registro original meta ruido.
DateTime dateOnly(DateTime date) => DateTime(date.year, date.month, date.day);

/// El lunes de la semana ISO que contiene [date] (1 = lunes).
DateTime mondayOfWeek(DateTime date) {
  final d = dateOnly(date);
  return d.subtract(Duration(days: d.weekday - 1));
}

/// Calcula las sesiones concretas de una lista de asignaciones que caen
/// dentro de `[rangeStart, rangeEndInclusive]` (los dos incluidos).
///
/// - Programa: la semana 1 de la plantilla arranca el lunes de la semana
///   en la que cae `start_date` (así que puede incluir días de esa misma
///   semana anteriores a la fecha exacta de inicio — se interpreta como
///   "el programa arranca esa semana", no como un corte a la mitad de la
///   semana 1). Se repite por cada semana hasta `duration_weeks`. Si el
///   cliente tiene un ajuste puntual para ese día (`assignment_overrides`),
///   se muestra la rutina de reemplazo, no la de la plantilla.
/// - Rutina suelta: se repite semana a semana en el mismo día de la
///   semana que `start_date`, sin fecha de fin (hasta que el entrenador
///   la reemplace o el cliente termine).
List<CalendarSession> sessionsInRange(
  List<CalendarAssignment> assignments, {
  required DateTime rangeStart,
  required DateTime rangeEndInclusive,
}) {
  final start = dateOnly(rangeStart);
  final end = dateOnly(rangeEndInclusive);
  final sessions = <CalendarSession>[];

  for (final assignment in assignments) {
    if (assignment.isProgram) {
      final durationWeeks = assignment.programDurationWeeks ?? 0;
      final week1Start = mondayOfWeek(assignment.startDate);
      final programEndExclusive = week1Start.add(
        Duration(days: durationWeeks * 7),
      );

      for (final slot in assignment.slots) {
        final date = week1Start.add(
          Duration(days: (slot.weekNumber - 1) * 7 + (slot.dayOfWeek.raw - 1)),
        );
        if (date.isBefore(start) || date.isAfter(end)) continue;
        if (!date.isBefore(programEndExclusive)) continue;

        final override = assignment.overridesByProgramRoutineId[slot.programRoutineId];
        sessions.add(
          CalendarSession(
            date: date,
            assignmentId: assignment.assignmentId,
            clientId: assignment.clientId,
            clientName: assignment.clientName,
            routineName: override ?? slot.routineName,
            isProgram: true,
            programName: assignment.programName,
            isCustomizedForClient: override != null,
          ),
        );
      }
    } else {
      final routineName = assignment.standaloneRoutineName;
      if (routineName == null) continue;

      final assignmentStart = dateOnly(assignment.startDate);
      final weekday = assignmentStart.weekday;
      // Primera ocurrencia de ese día de la semana dentro del rango
      // pedido (puede caer antes de rangeStart si el rango arranca a
      // mitad de semana).
      var date = start.add(Duration(days: (weekday - start.weekday) % 7));
      while (!date.isAfter(end)) {
        if (!date.isBefore(assignmentStart)) {
          sessions.add(
            CalendarSession(
              date: date,
              assignmentId: assignment.assignmentId,
              clientId: assignment.clientId,
              clientName: assignment.clientName,
              routineName: routineName,
              isProgram: false,
            ),
          );
        }
        date = date.add(const Duration(days: 7));
      }
    }
  }

  sessions.sort((a, b) => a.date.compareTo(b.date));
  return sessions;
}

/// Agrupa sesiones ya calculadas por día, en un mapa ordenado por fecha
/// ascendente (útil para pintar la grilla del calendario).
Map<DateTime, List<CalendarSession>> groupSessionsByDate(
  List<CalendarSession> sessions,
) {
  final map = <DateTime, List<CalendarSession>>{};
  for (final session in sessions) {
    map.putIfAbsent(session.date, () => []).add(session);
  }
  return map;
}

File: web/src/lib/calendar-logic.ts
/**
 * Cálculo de sesiones concretas de calendario a partir de las
 * asignaciones (programas y rutinas sueltas) de un entrenador — puerto
 * a TypeScript de `app/lib/features/trainer/domain/calendar_logic.dart`.
 *
 * Todas las fechas se manejan como llaves 'YYYY-MM-DD' (sin hora) para
 * no depender de la zona horaria del navegador en la aritmética.
 */

export interface CalendarProgramSlot {
  programRoutineId: string;
  weekNumber: number;
  dayOfWeek: number; // ISO: 1 = lunes ... 7 = domingo
  routineName: string;
}

export interface CalendarAssignment {
  assignmentId: string;
  clientId: string;
  clientName: string;
  startDate: string; // 'YYYY-MM-DD'
  isProgram: boolean;
  programName?: string | null;
  programDurationWeeks?: number | null;
  standaloneRoutineName?: string | null;
  slots: CalendarProgramSlot[];
  overridesByProgramRoutineId: Record<string, string>;
}

export interface CalendarSession {
  date: string; // 'YYYY-MM-DD'
  assignmentId: string;
  clientId: string;
  clientName: string;
  routineName: string;
  isProgram: boolean;
  programName?: string | null;
  isCustomizedForClient?: boolean;
}

function pad(n: number) {
  return String(n).padStart(2, "0");
}

export function toKey(y: number, m: number, d: number): string {
  return `${y}-${pad(m)}-${pad(d)}`;
}

export function todayKey(): string {
  const now = new Date();
  return toKey(now.getFullYear(), now.getMonth() + 1, now.getDate());
}

/** ISO weekday: 1 = lunes ... 7 = domingo. */
export function weekdayIso(key: string): number {
  const [y, m, d] = key.split("-").map(Number);
  const utcDay = new Date(Date.UTC(y, m - 1, d)).getUTCDay(); // 0 = domingo
  return utcDay === 0 ? 7 : utcDay;
}

export function addDays(key: string, days: number): string {
  const [y, m, d] = key.split("-").map(Number);
  const date = new Date(Date.UTC(y, m - 1, d));
  date.setUTCDate(date.getUTCDate() + days);
  return toKey(date.getUTCFullYear(), date.getUTCMonth() + 1, date.getUTCDate());
}

export function mondayOfWeek(key: string): string {
  return addDays(key, -(weekdayIso(key) - 1));
}

export function compareKeys(a: string, b: string): number {
  return a < b ? -1 : a > b ? 1 : 0;
}

/** Número de días entre dos llaves 'YYYY-MM-DD' (b - a). */
export function daysBetween(a: string, b: string): number {
  const [ay, am, ad] = a.split("-").map(Number);
  const [by, bm, bd] = b.split("-").map(Number);
  const da = Date.UTC(ay, am - 1, ad);
  const db = Date.UTC(by, bm - 1, bd);
  return Math.round((db - da) / 86400000);
}

/**
 * Sesiones concretas que caen dentro de [rangeStart, rangeEndInclusive]
 * (ambos incluidos).
 *
 * - Programa: la semana 1 arranca el lunes de la semana en la que cae
 *   `start_date`. El patrón de `duration_weeks` semanas se repite en
 *   ciclo indefinidamente — un programa de 1 semana se repite cada
 *   semana, uno de 2 semanas alterna semana A/B para siempre, etc. — y
 *   solo deja de aplicarse cuando el entrenador cambia la asignación del
 *   cliente. Si el cliente tiene un ajuste puntual
 *   (`assignment_overrides`) para esa casilla, se muestra la rutina de
 *   reemplazo (el ajuste aplica solo la primera vez que se repite esa
 *   semana del ciclo).
 * - Rutina suelta: se repite semana a semana en el mismo día de la
 *   semana que `start_date`, sin fecha de fin.
 */
export function sessionsInRange(
  assignments: CalendarAssignment[],
  rangeStart: string,
  rangeEndInclusive: string,
): CalendarSession[] {
  const sessions: CalendarSession[] = [];

  for (const assignment of assignments) {
    if (assignment.isProgram) {
      const durationWeeks = assignment.programDurationWeeks ?? 0;
      if (durationWeeks <= 0 || assignment.slots.length === 0) continue;
      const week1Start = mondayOfWeek(assignment.startDate);

      let date = rangeStart;
      while (compareKeys(date, rangeEndInclusive) <= 0) {
        if (compareKeys(date, week1Start) >= 0) {
          const weekIndex0 = Math.floor(daysBetween(week1Start, date) / 7);
          const cycleWeek = (weekIndex0 % durationWeeks) + 1;
          const dow = weekdayIso(date);

          for (const slot of assignment.slots) {
            if (slot.weekNumber !== cycleWeek || slot.dayOfWeek !== dow) continue;
            const override = assignment.overridesByProgramRoutineId[slot.programRoutineId];
            sessions.push({
              date,
              assignmentId: assignment.assignmentId,
              clientId: assignment.clientId,
              clientName: assignment.clientName,
              routineName: override ?? slot.routineName,
              isProgram: true,
              programName: assignment.programName,
              isCustomizedForClient: Boolean(override),
            });
          }
        }
        date = addDays(date, 1);
      }
    } else {
      const routineName = assignment.standaloneRoutineName;
      if (!routineName) continue;

      const assignmentStart = assignment.startDate;
      const weekday = weekdayIso(assignmentStart);
      let date = addDays(rangeStart, (weekday - weekdayIso(rangeStart) + 7) % 7);
      while (compareKeys(date, rangeEndInclusive) <= 0) {
        if (compareKeys(date, assignmentStart) >= 0) {
          sessions.push({
            date,
            assignmentId: assignment.assignmentId,
            clientId: assignment.clientId,
            clientName: assignment.clientName,
            routineName,
            isProgram: false,
          });
        }
        date = addDays(date, 7);
      }
    }
  }

  sessions.sort((a, b) => compareKeys(a.date, b.date));
  return sessions;
}

export function groupSessionsByDate(
  sessions: CalendarSession[],
): Map<string, CalendarSession[]> {
  const map = new Map<string, CalendarSession[]>();
  for (const session of sessions) {
    const list = map.get(session.date) ?? [];
    list.push(session);
    map.set(session.date, list);
  }
  return map;
}
